pyqasm/visitor.py

```diff
@@ -608,6 +611,71 @@ def _visit_gate_definition(self, definition: qasm3_ast.QuantumGateDefinition) ->
 
         return []
 
+    def _unroll_multiple_target_qubits(
+        self, operation: qasm3_ast.QuantumGate, gate_qubit_count: int
+    ) -> list[list[qasm3_ast.IndexedIdentifier]]:
+        """Unroll the complete list of all qubits that the given operation is applied to.
+           E.g. this maps 'cx q[0], q[1], q[2], q[3]' to [[q[0], q[1]], [q[2], q[3]]]
+
+        Args:
+            operation (qasm3_ast.QuantumGate): The gate to be applied.
+            gate_qubit_count (list[int]): The number of qubits that a single gate acts on.
+
+        Returns:
+            The list of all targets that the unrolled gate should act on.
+        """
+        op_qubits = self._get_op_bits(operation, self._global_qreg_size_map)
+        if len(op_qubits) % gate_qubit_count != 0:
+            raise_qasm3_error(
+                f"Invalid number of qubits {len(op_qubits)} for operation {operation.name.name}",
+                span=operation.span,
+            )
+        qubit_subsets = []
+        for i in range(0, len(op_qubits), gate_qubit_count):
+            # we apply the gate on the qubit subset linearly
+            qubit_subsets.append(op_qubits[i : i + gate_qubit_count])
+        return qubit_subsets
+
+    def _broadcast_gate_operation(
+        self, gate_function: Callable, all_targets: list[list[qasm3_ast.IndexedIdentifier]]
+    ) -> list[qasm3_ast.QuantumGate]:
+        """Broadcasts the application of a gate onto multiple sets of target qubits.
+
+        Args:
+            gate_function (callable): The gate that should be applied to multiple target qubits
+            all_targets (list[list[qasm3_ast.IndexedIdentifier]]):
+                The list of target of target qubits.
+                The length of this list indicates the number of time the gate is invoked.
+        Returns:
+            List of all executed gates.
+        """
+        result = []
+        for targets in all_targets:
+            result.extend(gate_function(*targets))
+        return result
+
+    def _update_qubit_depth(self, all_targets: list[list[qasm3_ast.IndexedIdentifier]]):
```

Comment:
I think for clarity we should name it something like `_update_qubit_depths_for_gate` as we have a lot of functions where depths are being modified. Eventually we wanna move towards one method where we handle this updation logic for all basic operations like `measure`, `reset`, `gate` etc.

Reply:
Fixed.